### ISSUE TITLE
JarFileResource incorrectly lists the contents of directories with spaces

### ISSUE THREAD

**Issue by gthelleloid** [eclipse/jetty.project#3983]
In [Issue 1539](https://github.com/eclipse/jetty.project/issues/1539), I noted JarFileResource failed to serve any resource with a space in its name that is located in a jar file. While my patch fixed that problem, it did not fix a related problem: a directory with a space in its name that is located in a jar file will not list any entries when queried for directory listing.

I have attached a zip file with two files:

1. test.jar, which contains a directory named "dir1" with a file "file1" and a directory named "dir 2" with a file "file2" (note the space in the second directory name).
2. JarDirWithSpaceTest.java, which shows that JarFileResource.list() returns the wrong number of files in "dir 2".

The Java file was compiled and run with a classpath including jetty-util-9.4.11.v20180605.jar.

The problem lies in JarFileResource.listEntries() in the line

String dir=_urlString.substring(_urlString.lastIndexOf("!/")+2);

Here, as in issue 1539, the substring value must be decoded before use. The fix is to change the line to 

String dir=URIUtil.decodePath(_urlString.substring(_urlString.lastIndexOf("!/")+2));

**Comment by joakime** [comment]
Do you have a concrete example?
Does this only apply to directory / path entries?

We have nasty JAR content tests already, guess we will just have to expand on these a bit.
See: https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DefaultServletTest.java
And its `jar-resource-odd.jar` - https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-servlet/src/test/resources/jar-resource-odd.jar

```
[jetty-9.4.x]$ ls -la *.jar
-rw-r--r-- 1 joakim joakim 2863 Jul 19 10:07 jar-resource-odd.jar
[jetty-9.4.x]$ jar -tvf jar-resource-odd.jar 
     0 Mon Apr 15 08:32:18 CDT 2019 META-INF/
    66 Mon Apr 15 08:32:18 CDT 2019 META-INF/MANIFEST.MF
     0 Wed Apr 10 13:43:20 CDT 2019 rez/
    17 Mon Apr 15 08:31:20 CDT 2019 rez/one
    12 Mon Apr 15 08:30:58 CDT 2019 rez/aaa
    12 Mon Apr 15 08:31:06 CDT 2019 rez/bbb
     0 Wed Apr 10 13:59:32 CDT 2019 rez/oddities/
    24 Wed Apr 10 13:47:34 CDT 2019 rez/oddities/;
    35 Wed Apr 10 13:49:02 CDT 2019 rez/oddities/#hashcode
    30 Wed Apr 10 13:49:18 CDT 2019 rez/oddities/index.html#fragment
    20 Wed Apr 10 13:59:32 CDT 2019 rez/oddities/other%2fkind%2Fof%2fslash
    40 Wed Apr 10 13:46:36 CDT 2019 rez/oddities/a file with a space
    53 Wed Apr 10 13:48:00 CDT 2019 rez/oddities/;" onmousedown="alert(document.location)"
    35 Wed Apr 10 13:46:22 CDT 2019 rez/oddities/some\slash\you\got\there
    12 Mon Apr 15 08:31:12 CDT 2019 rez/ccc
     0 Wed Apr 10 10:07:22 CDT 2019 rez/deep/
    30 Mon Apr 15 08:31:38 CDT 2019 rez/deep/xxx
    34 Mon Apr 15 08:31:50 CDT 2019 rez/deep/zzz
    17 Mon Apr 15 08:31:42 CDT 2019 rez/deep/yyy
```

**Comment by gthelleloid** [comment]
I'm sorry, I intended to attach a test case to my initial comment, but apparently I did not. I have now attached a zip file with slightly different contents than described before:

1. test.jar, which contains a directory named "dir1" with a file "file1" and a directory named "dir 2" with a file "file2" (note the space in the second directory name).
2. JarDirWithSpaceTest.java, which shows that JarFileResource.list() does not include "file2" when listing the contents of "dir 2" (but does include "file1" when listing the contents of "dir1").

The Java file was compiled and run with a classpath including jetty-util, jetty-io, jetty-http, and jetty-server (version 9.4.11.v20180605) and javax.servlet-api 3.1.0.

[test.zip](https://github.com/eclipse/jetty.project/files/3506869/test.zip)
